### PR TITLE
Feature: Conference calling restrictions - part 1

### PR DESCRIPTION
**Pull request by ohassine**
## What's new in this PR?
Jira: https://wearezeta.atlassian.net/browse/SQSERVICES-560
This  PR is a follow up of #3449 

It aims to:
- Restict client from making a conference call if he is not a part of paying team.
- Inform client once the team plan has been upgraded to a paying plan.

### Testing

- Tested manually

Team upgraded dialog | Upgrade now dialog
-- | --
Screenshot_1629751531 |  Screenshot_1629751562

#### APK
[Download build #3837](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3837/artifact/build/artifact/wire-dev-PR3458-3837.apk)
[Download build #3840](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3840/artifact/build/artifact/wire-dev-PR3458-3840.apk)
[Download build #3842](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3842/artifact/build/artifact/wire-dev-PR3458-3842.apk)
[Download build #3843](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3843/artifact/build/artifact/wire-dev-PR3458-3843.apk)
[Download build #3844](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3844/artifact/build/artifact/wire-dev-PR3458-3844.apk)
[Download build #3847](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3847/artifact/build/artifact/wire-dev-PR3458-3847.apk)

### CODE REVIEW

File: zmessaging/src/main/scala/com/waz/content/Preferences.scala [hunk 1]
```diff
@@ -441,4 +441,6 @@ object UserPreferences {
 
   lazy val FileSharingFeatureEnabled: PrefKey[Boolean] = PrefKey[Boolean]("file_sharing_feature_enabled", customDefault = true)
   lazy val ShouldInformFileSharingRestriction: PrefKey[Boolean] = PrefKey[Boolean]("should_inform_file_sharing_restriction", customDefault = false)
+  lazy val ConferenceCallingFeatureEnabled: PrefKey[Boolean] = PrefKey[Boolean]("conference_calling_feature_enabled", customDefault = true)
+  lazy val ShouldInformPlanUpgradedToEnterprise: PrefKey[Boolean] = PrefKey[Boolean]("should_inform_plan_upgraded_to_enterprise", customDefault = true)
```
Comment (johnxnguyen) [inline zmessaging/src/main/scala/com/waz/content/Preferences.scala:445]: I think having a custom default of `true` will cause the alert to be shown when first launch the app.

File: zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala [hunk 2]
```diff
@@ -61,12 +67,32 @@ class FeatureConfigsServiceImpl(syncHandler: FeatureConfigsSyncHandler,
...
 
+
+  override def updateConferenceCalling(): Future[Unit] =
+    for {
+      conferenceCalling <- syncHandler.fetchConferenceCalling()
+      _                 <- storeConferenceCallingConfig(conferenceCalling)
+    } yield ()
+
+  private def storeConferenceCallingConfig(conferenceCallingFeatureConfig: ConferenceCallingFeatureConfig): Future[Unit] = {
+    userPrefs (ConferenceCallingFeatureEnabled) := conferenceCallingFeatureConfig.isEnabled
```
Comment (johnxnguyen) [inline zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala:86]: The preference is set again just a few line later. Did you mean to delete this line?

File: zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala [hunk 3]
```diff
@@ -61,12 +67,32 @@ class FeatureConfigsServiceImpl(syncHandler: FeatureConfigsSyncHandler,
...
+    for {
+      existingValue <- userPrefs (ConferenceCallingFeatureEnabled).apply()
+      newValue      =  conferenceCallingFeatureConfig.isEnabled
+      _             <- userPrefs(ConferenceCallingFeatureEnabled) := newValue
+                    // Inform of plan upgraded.
+      _             <- if (!existingValue && newValue) userPrefs(ShouldInformPlanUpgradedToEnterprise) := true
+                       // Don't inform
+                       else if (newValue) userPrefs(ShouldInformPlanUpgradedToEnterprise) := false
+                       else Future.successful(())
```
Comment (johnxnguyen) [inline zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala:95]: Suppose conference calling is disabled. Then it becomes enabled twice in a row (for some reason). After the first change (from disabled to enabled) we will set the flag to inform the user that conference calling is now available. After the second update (from enabled to enabled), the `else if` branch will execute and so we will reset the inform flag.

If I understand correctly, if this all happens before the first alert dialog is shown, then no alert dialog will be shown at all. Is that what we want? Or should the `else if` condition be checking for `!newValue` instead?

**Review by johnxnguyen — COMMENTED** [review]
There are commits from the previous PR, did you mean to target this PR against `feature/conference_calling_restrictions_part_0`?

File: zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala (hunk 2 above)
Comment (ohassine) [reply zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala:86]: oh! yes to delete

File: zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala (hunk 3 above)
Comment (ohassine) [reply zmessaging/src/main/scala/com/waz/service/teams/FeatureConfigsService.scala:95]: Right! we should check `!newValue`

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**